rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -1552,61 +1552,98 @@ TEST(test_graph_cache, bad_arguments)
   }
 }
 
-TEST(test_graph_cache, get_writers_info_by_topic_maybe_fail)
+class TestGraphCache : public ::testing::Test
 {
-  RCUTILS_FAULT_INJECTION_TEST(
+public:
+  void SetUp()
   {
-    GraphCache graph_cache;
-    add_participants(graph_cache, {"participant1"});
+    add_participants(graph_cache_, {"participant1"});
     add_entities(
-      graph_cache,
+      graph_cache_,
     {
       // topic1
       {"reader1", "participant1", "topic1", "Str", true},
       {"writer1", "participant1", "topic1", "Str", false},
     });
     add_nodes(
-      graph_cache, {
+      graph_cache_, {
       {"participant1", "ns1", "node1"},
       {"participant1", "ns1", "node2"},
       {"participant1", "ns2", "node1"}});
+  }
 
-    rmw_topic_endpoint_info_array_t info = rmw_get_zero_initialized_topic_endpoint_info_array();
+protected:
+  GraphCache graph_cache_;
+};
 
+TEST_F(TestGraphCache, get_writers_info_by_topic_maybe_fail)
+{
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    rmw_topic_endpoint_info_array_t info = rmw_get_zero_initialized_topic_endpoint_info_array();
     rcutils_allocator_t allocator = rcutils_get_default_allocator();
-    rmw_ret_t ret = graph_cache.get_writers_info_by_topic(
+
+    rmw_ret_t ret = graph_cache_.get_writers_info_by_topic(
       "topic1",
       identity_demangle,
       &allocator,
       &info);
     if (RMW_RET_OK == ret) {
       ret = rmw_topic_endpoint_info_array_fini(&info, &allocator);
+      if (RMW_RET_OK != ret) {
+        // If fault injection causes fini to fail, it should work the second time.
+        EXPECT_EQ(RMW_RET_OK, rmw_topic_endpoint_info_array_fini(&info, &allocator));
+      }
     } else {
       rcutils_reset_error();
     }
+  });
+}
 
+TEST_F(TestGraphCache, get_node_names_maybe_fail)
+{
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
     rcutils_string_array_t names = rcutils_get_zero_initialized_string_array();
     rcutils_string_array_t namespaces = rcutils_get_zero_initialized_string_array();
     rcutils_string_array_t enclaves = rcutils_get_zero_initialized_string_array();
-    ret = graph_cache.get_node_names(&names, &namespaces, &enclaves, &allocator);
+    rcutils_allocator_t allocator = rcutils_get_default_allocator();
+    rmw_ret_t ret = graph_cache_.get_node_names(&names, &namespaces, &enclaves, &allocator);
     if (RMW_RET_OK == ret) {
-      ret = rcutils_string_array_fini(&names);
-      ret = rcutils_string_array_fini(&namespaces);
-      ret = rcutils_string_array_fini(&enclaves);
+      // rcutils_string_array_fini is not under test, so disable fault injection test here.
+      int64_t fault_injection_count = rcutils_fault_injection_get_count();
```

Comment:
@hidmic This is the only location here that would make sense to disable fault injection testing because `rcutils_string_array_fini` is out of scope for this test. Do you think it would make sense to submit a PR to rcutils to add in the disable fault injection testing macros?

Reply:
I think it's fine as-is for the time being. If it starts showing up more often, we can add a macro.